service/lib/dinstaller/dbus/y2dir/modules/Autologin.rb

```diff
@@ -0,0 +1,207 @@
+    end
+
+    def available
+      @available = supported? if @available.nil?
+      @available
+    end
+
+    # Read autologin settings
+    # @return used?
+    def Read
+      if SCR.Read(path(".target.size"), "/etc/sysconfig/displaymanager") == -1
+        @available = false
+        @user = ""
+        @used = false
+        return false
+      end
+
+      @available = supported?
+      @user = Convert.to_string(
+        SCR.Read(path(".sysconfig.displaymanager.DISPLAYMANAGER_AUTOLOGIN"))
+      )
+      @pw_less = Convert.to_string(
+        SCR.Read(
+          path(".sysconfig.displaymanager.DISPLAYMANAGER_PASSWORD_LESS_LOGIN")
+        )
+      ) == "yes"
+
+      @user = "" if @user == nil || @user == ""
+
+      @used = @user != ""
+      @used
+    end
+
+
+    # Write autologin settings
+    # @param [Boolean] write_only when true, suseconfig script will not be run
+    # @return written anything?
+    def Write(write_only)
+      return false if !available || !@modified
+
+      Builtins.y2milestone(
+        "writing user %1 for autologin; pw_less is %2",
+        @user,
+        @pw_less
+      )
+
+      SCR.Write(
+        path(".sysconfig.displaymanager.DISPLAYMANAGER_AUTOLOGIN"),
+        @user
+      )
+      SCR.Write(
+        path(".sysconfig.displaymanager.DISPLAYMANAGER_PASSWORD_LESS_LOGIN"),
+        @pw_less ? "yes" : "no"
+      )
+      SCR.Write(path(".sysconfig.displaymanager"), nil)
+
+      @modified = false
+      true
+    end
+
+    # Disable autologin
+    def Disable
+      @user = ""
+      @pw_less = false
+      @used = false
+      @modified = true
+
+      nil
+    end
+
+    # Wrapper for setting the 'used' variable
+    def Use(use)
+      if @used != use
+        @used = use
+        @modified = true
+      end
+
+      nil
+    end
+
+    # Disable autologin and write it (used probably for automatic
+    # disabling without asking)
+    # @param [Boolean] write_only when true, suseconfig script will not be run
+    # @return written anything?
+    def DisableAndWrite(write_only)
+      Disable()
+      Write(write_only)
+    end
+
+    # Ask if autologin should be disabled (and disable it in such case)
+    # @param [String] new The reason for disabling autologin (e.g. new set of users)
+    # @return Is autologin used?
+    def AskForDisabling(new)
+      # popup text (%1 is user name, %2 is additional info,
+      # like "Now LDAP was enabled")
+      question = Builtins.sformat(
+        _(
+          "The automatic login feature is enabled for user %1.\n" +
+            "%2\n" +
+            "Disable automatic login?"
+        ),
+        @user,
+        new
+      )
+
+      Disable() if @used && Popup.YesNo(question)
+      @used
+    end
+
+    # Check if autologin is supported with the currently selected or installed
+    # packages.
+    #
+    # @return Boolean
+    def supported?
+      supported = DInstaller::DBus::Clients::DInstaller.provisions_selected?(DISPLAY_MANAGERS).any?
```

Comment:
Are we copying the whole module only because this line?

Reply:
yes. It is how Y2DIR works. it is like self-update or LOAD_PATH in ruby. It prefers loading different file. For me it was easier then doing some dynamic monkey patching, as this needs to be done early due to some side effect when doing require/import. So this is safe way to do it without risking too early execution.